Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs

```diff
@@ -853,6 +854,71 @@ protected internal uint stateOffsetRelativeToDeviceRoot
         internal PrimitiveValue m_MinValue;
         internal PrimitiveValue m_MaxValue;
 
+        internal OptimizedControlDataType m_OptimizedControlDataType;
+        internal OptimizedControlDataType optimizedControlDataType => m_OptimizedControlDataType;
+        
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void CalculateOptimizedControlDataType()
+        {
+            // setting check need to be inline so we clear optimizations if setting is disabled after the fact
+            m_OptimizedControlDataType = InputSettings.optimizedControlsFeatureEnabled &&
+                                         this is IOptimizedControl optimizedControl
+                ? optimizedControl.CalculateOptimizedControlDataType()
+                : OptimizedControlDataType.NotOptimized;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void CalculateOptimizedControlDataTypeRecursively()
+        {
+            // Need to go depth-first because CalculateOptimizedControlDataType might depend on computed values of children
+            if (m_ChildCount > 0)
+                foreach (var inputControl in children)
+                    inputControl.CalculateOptimizedControlDataTypeRecursively();
+
+            CalculateOptimizedControlDataType();
+        }
+
+        // This function exists to warn users to start using ApplyParameterChanges for edge cases that were previously not intentionally worked,
+        // where control properties suddenly change underneath us without us anticipating that.
+        // This is mainly to AxisControl fields being public and capable of changing at any time even if we don't thing it's an anticipated usage pattern.
+        // Also it's not clear if InputControl.stateBlock.format can potentially change at any time, likely not.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
Maybe mark this as [Conditional("DEBUG")] instead of using #IF DEBUG, so there's zero cost at runtime. Just a method call but still.

Reply:
Fixed